Define more restrictively what it means that a cfd is closed

da-kami opened the pull request:
Fixes https://github.com/itchysats/itchysats/issues/1355

Previously we mostly used `is_final` as a guard, but it can take significant time until e.g. a collab close transaction gets confirmed which leads to other logic being triggered that is actually not possible anymore.
After a collab close transaction was published, other spend transactions will fail when publishing them, but since we don't handle errors in post-processing additional events will be appended so the `Cfd` was seemingly force-close even though it was already collaboratively closed.

---

Note that there are some other issues attached to this problem:

- We don't deal with in-mempool which is an old problem: #206 - this only changes the problem but not solves it. Instead of asserting that we *have* a transaction available (and assume that is was/will be published) we could assert on finding it in mempool.
- We are not dealing with post-processing failures. Effectively e.g. non-collab settlement *cannot* happen once the collab settlement tx is publsihed, but we store the event nontheless and then the post processing fails. Same for rollover (where the problem becomes visible in the UI more clearly, that's why we added this condition: #1315 )

da-kami committed:
Define more restrictively what it means that a cfd is closed

And use that to allow / disallow actions.
Previously we mostly used `is_final` as a guard, but it can take significant time until e.g. a collab close transaction gets confirmed which leads to other logic being triggered that is actually not possible anymore.
After a collab close transaction was published, other spend transactions will fail when publishing them, but since we don't handle errors in post-processing additional events will be appended so the `Cfd` was seemingly force-close even though it was already collaboratively closed.

This adds tests for scenarios involving rollover and settlement
Not all scenarios are covered, but the scenarios tested are the ones we had problems with in production.
Adding scenarios that involve asserting on a successful force close will require a test framework extension.

klochowicz commented on daemon/src/model/cfd.rs:
```diff
@@ -317,7 +317,7 @@ pub enum CfdEvent {
     /// This can happen in cases where we publish a settlement transaction while the lock
     /// transaction is still pending and they end up in the same block.
-    LockConfirmedAfterFinality,
+    LockConfirmedAfterClose,
```
would changing the existing event cause a lot of disruptions at this stage? perhaps we should append a new one instead of changing this one?

da-kami replied:
That's a very good point!

da-kami replied:
I opted for keeping the old name `LockConfirmedAfterFinality` which is not entirely correct, but I think acceptable. I extended the doc comment.

klochowicz commented on daemon/src/model/cfd.rs:
```diff
@@ -610,6 +608,10 @@ impl Cfd {
         self.settlement_proposal.is_some()
     }
 
+    fn is_in_non_collaborative_settlement(&self) -> bool {
```
`is_in_force_close()`? not sure whether we've reached full agreement on the naming yet, but last I checked we were leaning towards "close" and "force-close"

klochowicz commented on daemon/src/model/cfd.rs:
```diff
@@ -241,8 +241,8 @@ pub enum NoRolloverReason {
     #[error("Cannot roll over while CFD is in collaborative settlement")]
     InCollaborativeSettlement,
-    #[error("Cannot roll over when CFD is final")]
-    Final,
```
as per your commit message: an alternative way we could deal with appending events after "closing" cfd would be just treating the *first* closure as the one that got in (in other words, it could be handled in during rehydration phase - it doesn't need to look at the last event in all cases)

da-kami replied:
I think this goes against the paradigm of even-sourcing. We would start "ignoring" events. I think it's better to keep the even history correct. We could use it as a workaround to deal with `Cfds` that already are in a "wrong" state, but I think those should fix itself, once we get a confirmation on chain, because the confirmation will actually close it through finality.

klochowicz replied:
I think we can do both things eventually; trying not to get into a weird state in the first place is of course preferred as it makes our lives easier.
However, if we appended an event of e.g. closing a closed cfd, this could not be necessary bad, as long as this would lead to another event "closing failed" etc. Of course it goes outside of the scope of this PR, but it's possible :)

da-kami replied:
I'm happy to see work toward that in a follow up PR! I think it's an interesting idea

da-kami committed:
Allow collab settlement during rollover

Recently we made collab settlement and rollover mutually exclusive.
This is actually not entirely correct: Collaborative settlement can be triggered during a rollover, and should result in collaboratively settling. If a rollover finishes after a collaborative settlement then the rollover should just be dropped.
However, rollover should not possible during a collaborative settlement.
Note that the UI actually does not allow clicking collaborative settlement during a rollover at the moment, so we should almost never run into this anyway.

Additionally, we can run into scenarios where we become unable to collboratively settle, which should be avoided:
It can happen at the moment that we record `RolloverStarted` but then never record a completed event because e.g. a restart happens.
This should not happen, but if it happens and the the condition to not allow collaborative settlement "during" rollover is in place, we would not be able to collaboratively settle anymore. Note that in such a scenario rollover would also not work anymore, because we don't allow rollover if there is still an ongoing rollover, so we would eventually force close.

klochowicz commented on daemon/src/model/cfd.rs:
```diff
@@ -978,6 +996,9 @@ impl Cfd {
 
     /// Given an attestation, find and decrypt the relevant CET.
     pub fn decrypt_cet(self, attestation: &oracle::Attestation) -> Result<Option<Event>> {
+        // TODO: Should this be is_closed instead? i.e. should we *not* decrypt the CET if we have
```
it would make sense for me to change it, as trying to publish the transaction will fail anyway down the track?
Shall we add a test to cover this scenario, so we're explicit about desired handling of it? (not in this PR of course)

da-kami replied:
Yeah, in a later commit I actually did 😅 - should probably squash these.

klochowicz commented on daemon/src/model/cfd.rs:
```diff
@@ -654,8 +654,6 @@ impl Cfd {
 
     fn can_settle_collaboratively(&self) -> bool {
         !self.is_closed() && !self.commit_finality && !self.is_attested()
-            // Rollover and collaborative settlement are mutually exclusive, if we are currently rolling over we cannot settle
```
thanks for writing detailed reasoning for removing these lines! much appreciated!

klochowicz commented on daemon/src/model/cfd.rs:
```diff
@@ -2577,6 +2707,30 @@ mod tests {
             ]
         }
 
+        fn dummy_ongoing_collab_settlement(order_id: OrderId) -> Vec<Self> {
```
nit: I had a hard time reading what would "ongoing" mean - did you try say that it got started? maybe it's easier to say:
`apply_dummy_collab_settlement_started_event` or something similar?

da-kami replied:
yeah, the `Event` extension should not know about how this is used in the test so started makes more sense, because that's the only terminology the `Event` knows. 

From the test's perspective started might not be enough to understand why the test was written. I tried to convey "started but not finished" with ongoing... Not sure it actually helps.

It might actually be better to produce the dummy events more fine granular and then chain more together instead of bundling multiple events into one dummy scenario. We could give that a try in a refactor on top.

klochowicz commented on daemon/src/model/cfd.rs:
```diff
@@ -2389,6 +2390,135 @@ mod tests {
+            .with_quantity(quantity)
+            .with_opening_price(opening_price)
+            .with_leverage(leverage)
+            .dummy_open(dummy_event_id());
```
this boilerplate makes these tests not that readable.... I'll post a quick PR on top of it if you don't mind, I have something in mind for it 😀

da-kami replied:
definitely! If you have something in mind already that's very much appreciated. I wanted to get them done with what's currently there and then do a round of refactoring myself. If you already start / do with that's even better :)

klochowicz commented on daemon/src/model/cfd.rs:
```diff
@@ -2416,6 +2416,83 @@ mod tests {
+        let (cfd, _, _, _) = cfd
+            .with_lock(taker_keys, maker_keys)
+            .dummy_collab_settlement_taker(opening_price);
+
```
should we test with the ongoing collab settlement here? (or maybe in yet another test?) I have to admit still am pretty hazy in regards to the distinction "in collab settlement" vs "in ongoing collab settlement".

is this mapping right?
- "in collab settlement" - collab settlement proposed
- "in ongoing collab settlement" -  collab settlement got accepted and we're doing the protocol

da-kami replied:
The test below covers the `ongoing` case. I wanted to always cover both collab settlement finished (and we have the tx) and ongoing (where we have the settlement only). It's quite a lot of boilerplate which is annoying.

As for the mapping

> "in collab settlement" - collab settlement proposed
> "in ongoing collab settlement" - collab settlement got accepted and we're doing the protocol

that's not how I defined it, my definition is:

1. ongoing: we have a proposal but no collab settlement transaction yet (i.e. the settlement protocol is not finished yet)
2. completed: we have the collab settlement transaction (the settlement protocol completed successfully) but we don't know anything about finality yet
3. final: we reached finality of the transaction

We could document this somewhere, but not sure where...

klochowicz reviewed (approved):
I haven't found anything blocking, most of the comments I added are just questions - we can also clean it up a bit further later to minimise boilerplate, for which I'm happy to submit a PR :)

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -2371,6 +2390,212 @@ mod tests {
 
+    #[test]
+    fn given_collab_settlement_then_cannot_start_rollover() {
+        let quantity = Usd::new(dec!(10));
+        let leverage = Leverage::new(2).unwrap();
+        let opening_price = Price::new(dec!(10000)).unwrap();
```
These values are not important for the test right? Would be good to hide them then! Would `Cfd::taker_long().dummy_open()` be sufficient?

da-kami replied:
For tests that actually do a collaborative settlement (for either role) they are important because otherwise we run into different errors. For other tests I can try to remove them.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -2371,6 +2390,212 @@ mod tests {
+            .with_leverage(leverage)
+            .dummy_open(dummy_event_id());
+
+        let _ = assert!(cfd.start_rollover().is_ok());
```
The test name doesn't mention that we are testing the scenario of "rollover started - collab settlement started - rollover started".

Remove this?

da-kami replied:
it's more a sanity check that we are able to start rollover *before* we do whatever we do. But I can remove them. This goes towards "a test should only test one thing". Sometimes I find it useful to show to the reader that "at this point we can, then we do this, then we can't..." 

Besides that for some tests it was just copy paste and completely irrelevant for the test 😅
I removed all of them because I don't think they add that much value alltogether. We should potentially test that we can start rollover in separate tests.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -2371,6 +2390,212 @@ mod tests {
+
+        let err = cfd.start_rollover().unwrap_err();
+        let no_rollover_reason = err.downcast::<NoRolloverReason>().unwrap();
+        assert_eq!(no_rollover_reason, NoRolloverReason::Closed);
```
Nit: Could do better use of whitespace here:

```suggestion
	// act
        let (cfd, _, _, _) = cfd
            .with_lock(taker_keys, maker_keys)
            .dummy_collab_settlement_taker(opening_price);
        let result = cfd.start_rollover();
        
        // assert
        let no_rollover_reason = result.unwrap_err().downcast::<NoRolloverReason>().unwrap();
        assert_eq!(no_rollover_reason, NoRolloverReason::Closed);
```

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -2373,6 +2390,219 @@ mod tests {
+        let leverage = Leverage::new(2).unwrap();
+        let opening_price = Price::new(dec!(10000)).unwrap();
+
+        let taker_keys = crate::keypair::new(&mut rand::thread_rng());
+        let maker_keys = crate::keypair::new(&mut rand::thread_rng());
```
Would be nice to get rid of the keys on this level if we don't need them for anything else beyond the `with_lock` invocation.

da-kami replied:
yeah, but for *most* cases we need them outside because they have to be the same for taker and maker. I think this should go away once we get your changes for a simplified lock in that we don't build collaboratively.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -2373,6 +2390,219 @@ mod tests {
+    /// over the rollover. Upon finishing the rollover we fail because the cfd was already
+    /// settled.
+    #[test]
+    fn given_collab_settlement_then_cannot_finish_rollover() {
```
```suggestion
    fn given_collab_settlement_finished_then_cannot_finish_rollover() {
```

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -2373,6 +2390,219 @@ mod tests {
+                OrderId::default(),
+                Dlc::dummy(None),
+                FundingFee::dummy(),
+            ))
```
This is the "act" part of the test. Would be nice to indicate this with good use of whitespace!

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -2373,6 +2390,219 @@ mod tests {
+            .with_leverage(leverage)
+            .dummy_open(dummy_event_id());
+
+        let _ = assert!(cfd.start_rollover().is_ok(), "Could not start rollover");
```
I don't think this should be here according to the test desc.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -2373,6 +2390,219 @@ mod tests {
+        let no_rollover_reason = err.downcast::<NoRolloverReason>().unwrap();
+        assert_eq!(
+            no_rollover_reason,
+            NoRolloverReason::InCollaborativeSettlement
```
This is a completely separate test, isn't it? (the maker side of things)

da-kami replied:
Yes, but the same test context. I could split it up in `_taker` and `_maker` test but that goes towards the same discussion of being able to group.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -2373,6 +2390,219 @@ mod tests {
+        let (taker_long, proposal, taker_sig, _) = taker_long
+            .with_lock(taker_keys, maker_keys)
+            .dummy_start_rollover()
+            .dummy_collab_settlement_taker(opening_price);
```
It is pretty confusing that the "act" part of this test is chained together with the "arrange". Can we split this please?

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -2373,6 +2390,219 @@ mod tests {
+
+    #[test]
+    fn given_collab_settlement_then_cannot_force_close() {
+        let quantity = Usd::new(dec!(10));
+        let leverage = Leverage::new(2).unwrap();
+        let opening_price = Price::new(dec!(10000)).unwrap();
+
+        let taker_keys = crate::keypair::new(&mut rand::thread_rng());
+        let maker_keys = crate::keypair::new(&mut rand::thread_rng());
```
So. Much. Noise. :grimacing:

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -2373,6 +2390,219 @@ mod tests {
+        assert!(
+            cfd.manual_commit_to_blockchain().is_err(),
+            "Manual commit to blockchain did not error"
+        );
```
Why not `unwrap_err` and assert on the error, i.e. call `to_string` on it and assert on the reason?

thomaseizinger reviewed (dismissed):
Lots of noise to be removed but overall good effort, thanks!

I think there are a couple of occasions where we start a rollover although it shouldn't be part of the test. Would like to discuss that otherwise ACK.
